[SPARK-4598][WebUI]Task table pagination for the Stage page

zsxwing opened the pull request:
This PR adds pagination for the task table to solve the scalability issue of the stage page. Here is the initial screenshot:
<img width="1347" alt="pagination" src="https://cloud.githubusercontent.com/assets/1000778/8679669/9e63863c-2a8e-11e5-94e4-994febcd6717.png">
The task table only shows 100 tasks. There is a page navigation above the table. Users can click the page navigation or type the page number to jump to another page. The table can be sorted by clicking the headers. However, unlike previous implementation, the sorting work is done in the server now. So clicking a table column to sort needs to refresh the web page.

zsxwing committed:
Task table pagination for the Stage page

AmplabJenkins commented:
Build triggered.

zsxwing commented:
I will clean this PR soon. @pwendell any thoughts about the screenshot?

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #37239 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37239/consoleFull) for   PR 7399 at commit [`db6c859`](https://github.com/apache/spark/commit/db6c8596fef51929fadbb4a931ad700336ebcd72).

SparkQA commented:
[Test build #37239 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37239/console) for   PR 7399 at commit [`db6c859`](https://github.com/apache/spark/commit/db6c8596fef51929fadbb4a931ad700336ebcd72).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

andrewor14 commented:
Yes!!

XuTingjun commented:
This table can sort globally by any field?

zsxwing commented:
> Can the table sort globally by any field?

Yes. It will sort the data by the column and slice the corresponding data for a page in the server.

zsxwing committed:
Merge branch 'master' into task-table-pagination

zsxwing committed:
Add comments and fix the style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37366/consoleFull) for   PR 7399 at commit [`d9285f0`](https://github.com/apache/spark/commit/d9285f0a33b556dd1978c45e3e0c3963d65bceb3).

zsxwing commented:
Here is the rule for page navigation.
* If the totalPages is 1, the page navigation will be empty
* If the totalPages is more than 1, it will create a page navigation including a group of page numbers (10 pages) and a form to submit the page number.

Here are some examples of the page navigation:
```
<< < 11 12 13* 14 15 16 17 18 19 20 > >>
   
This is the first group, so "<<" is hidden.
< 1 2* 3 4 5 6 7 8 9 10 > >>
   
This is the first group and the first page, so "<<" and "<" are hidden.
1 2* 3 4 5 6 7 8 9 10 > >>
   
Assume totalPages is 19. This is the last group, so ">>" is hidden.
<< < 11 12 13* 14 15 16 17 18 19 >
   
Assume totalPages is 19. This is the last group and the last page, so ">>" and ">" are hidden.
<< < 11 12 13 14 15 16 17 18 19*
   
* means the current page number
<< means jumping to the first page of the previous group.
< means jumping to the previous page.
>> means jumping to the first page of the next group.
> means jumping to the next page.
```

And a screenshot for 100k tasks:

<img width="1356" alt="pagination-100k" src="https://cloud.githubusercontent.com/assets/1000778/8702111/545e57aa-2b48-11e5-82c0-e488617d31a2.png">

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -231,52 +241,25 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
         accumulableRow,
         accumulables.values.toSeq)
 
-      val taskHeadersAndCssClasses: Seq[(String, String)] =
```
`taskHeadersAndCssClasses` is moved to `TaskPagedTable.headers`

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -853,32 +1280,4 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
-    }
-  }
-
-  private def getSchedulerDelay(info: TaskInfo, metrics: TaskMetrics, currentTime: Long): Long = {
```
It's moved to object StagePage.getSchedulerDelay

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -853,32 +1280,4 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
     <td>{errorSummary}{details}</td>
   }
-
-  private def getGettingResultTime(info: TaskInfo, currentTime: Long): Long = {
```
It's moved to object StagePage.getGettingResultTime

zsxwing commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
     </script>
   }
 
-  def taskRow(
```
This method is split to two methods: `TaskDataSource.taskRow` and `TaskPagedTable.row`.

SparkQA commented:
[Test build #37366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37366/console) for   PR 7399 at commit [`d9285f0`](https://github.com/apache/spark/commit/d9285f0a33b556dd1978c45e3e0c3963d65bceb3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

carsonwang commented:
The REST API provides all the task information in JSON format. It seems it is easier to enable pagination on client side by consuming the JSON data. I was actually doing something like this. Is there any reason not consuming the JSON data on the UI?

zsxwing commented:
> The REST API provides all the task information in JSON format. It seems it is easier to enable pagination on client side by consuming the JSON data. I was actually doing something like this. Is there any reason not consuming the JSON data on the UI?

I chose this way because the current Spark UI doesn't use the REST API. In addition, we cannot write unit tests for JavaScript now, so it's better to use Scala.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -231,52 +241,25 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+        stageData.hasBytesSpilled,
+        currentTime,
+        page = taskPage,
+        pageSize = 100, // Show 100 tasks at most in the table
```
Can we make this something higher? Maybe 200?

pwendell replied:
Actually on second though I prefer 100.

pwendell commented:
I took a look at the UI and I like it. I did have two thoughts though:

1. Can we make it so the headers, when hovering, show the link cursor? Otherwise it is not clear they are clickable.
2. Can we also make the page links show the link cursor when hovering?
3. To the left of the page selector, can you have the word "Page:" otherwise it looks like those are task numbers, since the header says "Tasks".

carsonwang commented on core/src/main/scala/org/apache/spark/ui/PagedTable.scala:
```diff
@@ -0,0 +1,221 @@
+   * < 1 2* 3 4 5 6 7 8 9 10 > >>
+   *
+   * This is the first group and the first page, so "<<" and "<" are hidden.
+   * 1 2* 3 4 5 6 7 8 9 10 > >>
```
Is this 1* 2 3 4 ?

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+  private def taskRow(taskData: TaskUIData): TaskTableRowData = {
+    val TaskUIData(info, metrics, errorMessage) = taskData
+    val duration = if (info.status == "RUNNING") info.timeRunning(currentTime)
+    else metrics.map(_.executorRunTime).getOrElse(1L)
```
Indentation

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+    val duration = if (info.status == "RUNNING") info.timeRunning(currentTime)
+    else metrics.map(_.executorRunTime).getOrElse(1L)
+    val formatDuration = if (info.status == "RUNNING") UIUtils.formatDuration(duration)
+    else metrics.map(m => UIUtils.formatDuration(m.executorRunTime)).getOrElse("")
```
Indentation

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+
+  override def tableId: String = ""
+
+  override  def tableCssClass: String = "table table-bordered table-condensed table-striped"
```
One more space here

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+  )
+
+  override def pageLink(page: Int): String = {
+    s"${basePath}&task.page=$page&task.sort=${sortColumn}&task.desc=${desc}"
```
Do we need encode the sortColumn here as well?

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+          }
+        } else {
+          throw new IllegalArgumentException(
+            "Cannot sort by Input Size / Records because of no outputs")
```
"Cannot sort by Output Size" here

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+          }
+        } else {
+          throw new IllegalArgumentException(
+            "Cannot sort by Input Size / Records because of no shuffle reads")
```
"Cannot sort by Shuffle Read Blocked Time"

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+        if (hasInput) {
+          new Ordering[TaskTableRowData] {
+            override def compare(x: TaskTableRowData, y: TaskTableRowData): Int =
+              Ordering.String.compare(x.input.get.inputSortable, y.input.get.inputSortable)
```
If we compare the values by string, it sorts the values in alphabetical order. We actually want to sort them by Long.

carsonwang commented on core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:
```diff
@@ -672,164 +658,605 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+        if (hasOutput) {
+          new Ordering[TaskTableRowData] {
+            override def compare(x: TaskTableRowData, y: TaskTableRowData): Int =
+              Ordering.String.compare(x.output.get.outputSortable, y.output.get.outputSortable)
```
The same as above. There are also other places that sort by String.

carsonwang commented:
@zsxwing , I just left some minor comments. 
Another problem: If a user selects to show additional metrics and sort by one metric like SchedulerDelay, the server will return a page without showing the additional metrics. We may need remember what additional metrics the user selects, and show these metrics when returning the page.

zsxwing committed:
Address Carson's comments

zsxwing committed:
Show the link cursor when hovering for headers and page links and other minor fix

zsxwing committed:
Use localStorage to remember the user's actions and replay them when loading the page

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
@carsonwang Thank you for reviewing this. I fixed the typos and code styles as well as the wrong usages of `Ordering.String`.

> Another problem: If a user selects to show additional metrics and sort by one metric like SchedulerDelay, the server will return a page without showing the additional metrics. We may need remember what additional metrics the user selects, and show these metrics when returning the page.

Good point. I used `localStorage` to remember them and replay them when loading the page.

SparkQA commented:
[Test build #37636 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37636/consoleFull) for   PR 7399 at commit [`b123f67`](https://github.com/apache/spark/commit/b123f67a7bdaa184ad24acd9b30fb4053bd13e05).

zsxwing commented:
@pwendell now when the user moves to the headers or page links, it shows the link cursor. And added "Page: ", too. Here is the new screenshot:

<img width="1338" alt="pagination-100k-new" src="https://cloud.githubusercontent.com/assets/1000778/8749833/a303c388-2cd6-11e5-8e85-9402452ad7ee.png">

SparkQA commented:
[Test build #37636 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37636/console) for   PR 7399 at commit [`b123f67`](https://github.com/apache/spark/commit/b123f67a7bdaa184ad24acd9b30fb4053bd13e05).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
@zsxwing I played around with this patch offline a little bit and it works great! There are two feature requests from my side:
- What if I have 500 tasks, and I just want to see everything on the timeline? Right now there's no way to do that. Perhaps there should be a `show all tasks` button somewhere or something.
- When I sort the column, it refreshes and jumps to the top of the page. This is very annoying because in old Spark versions I'm used to just clicking on the columns one after another. It would be good if the browser remembers the scroll position of the page before the refresh. E.g. something like http://stackoverflow.com/questions/17642872/refresh-page-and-keep-scroll-position

andrewor14 commented on core/src/main/resources/org/apache/spark/ui/static/additional-metrics.js:
```diff
@@ -26,17 +29,31 @@ $(function() {
     $('input[type="checkbox"]').click(function() {
-        var column = "table ." + $(this).attr("name");
+        var name = $(this).attr("name")
+        var column = "table ." + name;
```
is the space between `table` and `.` here intentional? I notice this is also in the old code

zsxwing replied:
Yes. I think so. `.` is a css class selector. E.g., if `name` is `deserialization_time`, it will select all cells in the table like `<td sorttable_customkey="0" class="deserialization_time">`.

andrewor14 commented on core/src/main/resources/org/apache/spark/ui/static/additional-metrics.js:
```diff
@@ -26,17 +29,31 @@ $(function() {
+        window.localStorage.setItem("expand-additional-metrics", "" + status);
     });
 
+    if (window.localStorage.getItem("expand-additional-metrics") == "true") {
```
minor, but we probably don't need this check; if it's already false then it doesn't do anything if we set it to false again. Same in L74 and in other files.

andrewor14 commented on core/src/main/resources/org/apache/spark/ui/static/spark-dag-viz.js:
```diff
@@ -93,8 +96,24 @@ function toggleDagViz(forJob) {
 }
 
+$(function (){
+  if (window.localStorage.getItem("expand-dag-viz-arrow-false") == "true") {
```
it's not intuitive what `expand-dag-viz-arrow-false` means. I would expect something like `expand-dag-viz-arrow-stage-page` or something.

andrewor14 commented on core/src/main/scala/org/apache/spark/ui/PagedTable.scala:
```diff
@@ -0,0 +1,222 @@
+  /**
+   * Slice the data for this page
+   */
+  def pageData: PageData[T] = {
```
can this be a val?

zsxwing replied:
No. It uses `data`. So it cannot be `val` because `data` is null in this class's constructor.

andrewor14 replied:
ah, I see. But it can be a lazy val right?

andrewor14 commented on core/src/main/scala/org/apache/spark/ui/PagedTable.scala:
```diff
@@ -0,0 +1,222 @@
+   * < means jumping to the previous page.
+   * >> means jumping to the first page of the next group.
+   * > means jumping to the next page.
+   * }}}
```
nice. This java doc is very clear

andrewor14 commented:
By the way, one more thing is that when I put in Page 111 when there are only 100 pages, it gives me the following error. We could have a more user-friendly error. This is minor and does not have to be fixed in this patch.

<img width="417" alt="screen shot 2015-07-17 at 1 34 40 pm" src="https://cloud.githubusercontent.com/assets/2133137/8757328/30dc95de-2c8e-11e5-8d6d-bc9b70948b5d.png">

zsxwing commented:
> When I sort the column, it refreshes and jumps to the top of the page. This is very annoying because in old Spark versions I'm used to just clicking on the columns one after another. It would be good if the browser remembers the scroll position of the page before the refresh. E.g. something like http://stackoverflow.com/questions/17642872/refresh-page-and-keep-scroll-position

This is cool. I will add it.

> What if I have 500 tasks, and I just want to see everything on the timeline? I know it's not going to crash the driver / the browser because it's not actually that many tasks, but right now there's no way to do that because we always display 100 tasks at a time. Perhaps there should be a show all tasks button somewhere or something.
> 
> In the future it would be good to be able to customize how many tasks to show per page on the UI itself. A use case I can think of is: I want to see the timeline for the top 1000 tasks with highest scheduler delay. If it's a lot of work we should just add this separately and hardcode 100 for now.

It's easy to add a text field for users to set the page size. It should solve these two problems.

zsxwing commented:
> By the way, one more thing is that when I put in Page 111 and press Go when there are only 100 pages, it gives me the following error. We could have a more user-friendly error. This is minor and does not have to be fixed in this patch.

Maybe we should add an error page for the Spark UI? Now this is the default Jetty error page.

zsxwing committed:
Use expand-dag-viz-arrow-job and expand-dag-viz-arrow-stage instead of expand-dag-viz-arrow-true and expand-dag-viz-arrow-false

andrewor14 commented:
> Maybe we should add an error page for the Spark UI? Now this is the default Jetty error page.

I was thinking just displaying the normal stage page, but replace the task table with an informative message, something like "Page 111 is out of range. Please select a page number between 1 and 100." We probably don't need to add an error page.

zsxwing committed:
Scroll down to the tasks table if the url contains any sort column

zsxwing committed:
Display user-friendly error messages

zsxwing committed:
Add a text field to let users fill the page size

zsxwing commented:
@andrewor14 the latest commits addressed your comments. Here is the new screenshot:

<img width="1350" alt="pagination-100k-200" src="https://cloud.githubusercontent.com/assets/1000778/8766577/4ef33044-2e70-11e5-9150-d90fe37873c5.png">

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37774 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37774/consoleFull) for   PR 7399 at commit [`c2f7f39`](https://github.com/apache/spark/commit/c2f7f39fc0dee690ae02b2f5a74f8d082d7b2d98).

SparkQA commented:
[Test build #37774 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37774/console) for   PR 7399 at commit [`c2f7f39`](https://github.com/apache/spark/commit/c2f7f39fc0dee690ae02b2f5a74f8d082d7b2d98).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Hi @zsxwing I tried this out locally. It looks like there are a few usability issues:
- When I put in page 111 / 100, I see the following, which is good. However, all the buttons are now missing, and I can't "select a page number between 1 and 100" as the error message suggested me to do.
<img width="1190" alt="screen shot 2015-07-19 at 12 28 37 pm" src="https://cloud.githubusercontent.com/assets/2133137/8767442/e53e9282-2e11-11e5-9dc4-140854fb1da8.png">
- Also, if you look at the screenshot, there is no space between the error message and the bottom of the page. It would be good to add some padding there.
- When I tried to increase the number of tasks shown per page, it gives me the above error message. For instance, I was previously on page 99 / 100, and then I increased the number of tasks displayed from 100 to 200, then I got an error telling me to select a number between 1 and 50. I think we should just clear the other variables (ordering, current page number etc.) when we change this variable.

zsxwing committed:
Reset page to 1 if the user changes the page size

zsxwing committed:
Add extra space for the error message

zsxwing committed:
Display the page navigation when the page number is out of range

zsxwing commented:
@andrewor14 Here is the new screenshot for the error message:

<img width="1350" alt="screen shot 2015-07-20 at 2 36 48 pm" src="https://cloud.githubusercontent.com/assets/1000778/8770962/0ac43d8c-2eed-11e5-90e2-2fef6949d9a4.png">

And if the user updates the page size, the page number will be reset to 1.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/35/consoleFull) for   PR 7399 at commit [`144f513`](https://github.com/apache/spark/commit/144f5133d8c40cae37af62cb03bac2a7a1443ce9).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37816 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37816/consoleFull) for   PR 7399 at commit [`144f513`](https://github.com/apache/spark/commit/144f5133d8c40cae37af62cb03bac2a7a1443ce9).

SparkQA commented:
[Test build #35 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/35/console) for   PR 7399 at commit [`144f513`](https://github.com/apache/spark/commit/144f5133d8c40cae37af62cb03bac2a7a1443ce9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #37816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37816/console) for   PR 7399 at commit [`144f513`](https://github.com/apache/spark/commit/144f5133d8c40cae37af62cb03bac2a7a1443ce9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Looks great!! I'll try this out later today again.

andrewor14 commented:
LGTM, I'm merging this into master. Thanks @zsxwing.
